emote/algorithms/amp.py

```diff
@@ -0,0 +1,184 @@
+                    the discriminator is trained to classify data from this batch as
+                    positive samples
+                policy_batch (dict): a batch of data from the RL buffer. the discriminator
+                    is trained to classify data from this batch as negative samples.
+            Returns:
+                loss (Tensor): the loss tensor
+        """
+        imitation_batch_size = imitation_batch["batch_size"]
+        policy_data_batch_size = policy_batch["batch_size"]
+
+        pos_obs: Tensor = imitation_batch["observation"][self._obs_key]
+        pos_next_obs: Tensor = imitation_batch["next_observation"][self._obs_key]
+        neg_obs: Tensor = policy_batch["observation"][self._obs_key]
+        neg_next_obs: Tensor = policy_batch["next_observation"][self._obs_key]
+
+        pos_obs = self._imitation_state_map_function(pos_obs)
+        pos_next_obs = self._imitation_state_map_function(pos_next_obs)
+        pos_input = torch.cat([pos_obs, pos_next_obs], dim=-1)
+        pos_input.requires_grad_(True)
+
+        neg_obs = self._policy_state_map_function(neg_obs)
+        neg_next_obs = self._policy_state_map_function(neg_next_obs)
+        neg_input = torch.cat([neg_obs, neg_next_obs], dim=-1)
+
+        pos_output = self._discriminator(pos_input)
+        neg_output = self._discriminator(neg_input)
+        assert pos_output.shape == (imitation_batch_size, 1)
+        assert neg_output.shape == (policy_data_batch_size, 1)
+
+        pos_loss = torch.mean(torch.square(pos_output - 1.0))  # Positive samples should label to 1.
+        neg_loss = torch.mean(
+            torch.square(neg_output + 1.0)
+        )  # Negative samples should label to -1.
+
+        grad_penalty_loss = self._grad_loss_weight * gradient_loss_function(pos_output, pos_input)
+
+        loss = pos_loss + neg_loss + grad_penalty_loss
+
+        self.log_scalar("amp/loss/pos_discrimination_loss", pos_loss)
+        self.log_scalar("amp/loss/neg_discrimination_loss", neg_loss)
+        self.log_scalar("amp/loss/grad_loss", grad_penalty_loss)
+        self.log_scalar("amp/loss/total", loss)
+        self.log_scalar("amp/predict/positive_samples_mean", torch.mean(pos_output))
+        self.log_scalar("amp/predict/positive_samples_std", torch.std(pos_output))
+        self.log_scalar("amp/predict/negative_samples_mean", torch.mean(neg_output))
+        self.log_scalar("amp/predict/negative_samples_std", torch.std(neg_output))
+
+        return loss
+
+
+class AMPReward(Callback, LoggingMixin):
+    """Adversarial rewarding with AMP"""
+
+    def __init__(
+        self,
+        discriminator: nn.Module,
+        state_map_fn: Callable[[Tensor], Tensor],
+        style_reward_weight: float,
+        rollout_length: int,
+        observation_key: str,
+        data_group: str,
+    ):
+        super().__init__()
+        self._discriminator = discriminator
+        self._order = 0
```

Comment:
Can this become a problem when paired with SAC? If QTarget runs before this callback it will miss the amp reward. Perhaps order just be an input parameter?

Reply:
What happens if QTarget also has order=0?

Reply:
Alright, I think this is fine then for this PR, users will have to assert that it runs before QTarget (or any other relevant callback).

I think I prefer to have the constructors take in an `order` parameter rather than having a `set_order` method. The callbacks are sorted by their order only once when given to the `Trainer`, so if you would call `set_order` at a later time, it would not have any effect on the actual training. By only exposing it during construction makes it clear that this is a "set once, never change" thing. A `get_order` method sounds nice though, we could make it a `@property`.

Anyway, that sounds like a different PR, approving this, nice stuff!